### ISSUE TITLE
Flambda 2 classic mode unboxing shortcomings discovered in PR1470

### ISSUE THREAD

**Issue by mshinwell** [ocaml-flambda/flambda-backend#1491]
That PR is going to disable two testcases in classic mode.

@Gbury investigated and wrote:

the first one is because we currently do not unbox continuations parameters in to_cmm: with @lthls  we have a plan that would allow us to do that (basically: analyse things during from_lambda, and then during to_cmm unbox the continuation parameters that were marked by from_lambda).

the second one is due to local allocs: currently local allocs have all kinds of effects and coeffects (mainly to prevent them from being moved outside of regions) and that is preventing to_cmm from doing anything with them. To unbox them properly, we'd need a proper tracking system for local regions in to_cmm (and probably simplify), so that we can give local allocs the proper effects and coeffects (or even treat them the same as regular allocs)

**Comment by Gbury** [comment]
More info about the second one, consider the following code:
```ocaml
let f x y =
  let z = x +. y in
  let t = x *. y in
  t = z
```

We should be able to eliminate all allocations in such simple code, and we do so in `-O3`. However, in classic mode, we rely on cmm optimizations to unbox things, and this doesn't work on local allocations, and thus we get the following cmm code:
```
(function{/home/guigui/tmp/local_classic_float.ml:2,6-59}
 camlLocal_classic_float__f_0_0_code (x32/285: val y33/286: val)
 (let
   (region34/287 (beginregion)
    z39/291
      (alloc_local{/home/guigui/tmp/local_classic_float.ml:3,10-16} 1789
        (+f (load float64 x32/285) (load float64 y33/286)))
    t43/295
      (alloc_local{/home/guigui/tmp/local_classic_float.ml:4,10-16} 1789
        (*f (load float64 x32/285) (load float64 y33/286)))
    region_return35/300
      (+ (<< (==f (load float64 t43/295) (load float64 z39/291)) 1) 1)
    unit48/301 (seq (endregion region34/287) 1))
   region_return35/300))
```

This is an example where silently and automatically turning things into local allocations is strictly worse (at least until we manage to better handle local allocs in to_cmm).

**Comment by lpw25** [comment]
There was a bug preventing local allocations from getting unboxed in `cmmgen` which I've fixed in #1710. That still doesn't fix the example on classic mode though, so there is something else going on there.

**Comment by lthls** [comment]
The unboxing for local boxed numbers in Cmmgen transforms these local allocations into heap allocations, so it can then move the allocation wherever it wants to. If we're fine with that, we could 'fix' classic mode unboxing by always allocating boxed numbers on the heap.
What we don't have currently in classic mode is an analysis that tells us what should be unboxed and what should not, so we can't make subtle decisions.

**Comment by Gbury** [comment]
It's very clear (at least for me) what is hapening in the case of the classic mode and prevents unboxing from working:
- flambda2 gives both arbitrary effects and coeffects to local allocations (including allocation of boxed numbers)
- this is useful to ensure that local allocations do not get moved across region boundaries (which also have effects and coeffects)
- however, that makes to_cmm very limited in its ability to re-order things since it relies on the effects and coeffects, and therefore it would be illegal to even switch two local allocations (at least according to their effects and co-effects). Add to that the fact that, by default, substitution of effect-ful expressions is disabled in to_cmm (there's a flag for that iirc).

The proper solution for that would be to stop encoding the lifetime restrictions of local values with respect to their regions into the effects and coeffect system, since it was not designed to do that, and that transformation loses quite a lot of information.

**Comment by lpw25** [comment]
> The unboxing for local boxed numbers in Cmmgen transforms these local allocations into heap allocations

What makes you say that? It looks like it preserves them as local allocations to me.

> flambda2 gives both arbitrary effects and coeffects to local allocations

Shouldn't they only be marked as coeffectful? What requires them to be marked as effectful?

Besides, simple examples like yours don't require them to be moved across any operations. The old unanf pass manages to handle these just fine.

**Comment by Gbury** [comment]
> The old unanf pass manages to handle these just fine.

For the un_anf pass: I don't know how local allocations are handled in flambda1 (in particular what are their effects and coeffects in flambda1), and I don't think it really matters here ? The point is rather how they are handled in flambda2, and with the current scheme of using effects and coeffects, there isn't much more that can be done without breaking the semantics of effects and coeffects.

> Besides, simple examples like yours don't require them to be moved across any operations.

My point is that, because of the use of effects, any other local allocations is an operation across which local allocations cannot be moved, so the unboxing would simply stop working if you changed the order of let-definitions. Basically, using effects and coeffects is very brittle because these were not designed to represent the restrictions that must be enforced for local allocations.

> Shouldn't they only be marked as coeffectful? What requires them to be marked as effectful?

Even if only given co-effects, using effects and coeffects to encode the lifetime restrictions on local allocations has a limited expressiveness. We're currently hitting that limitation in the context of to_cmm unboxing specifically, but I'm sure that there are other examples where it hinders simplifications. So rather than discussing how to make the encoding works, I'd prefer if we could talk about a proper system for tracking where local allocations are allowed to be moved, that doesn't rely on effects and coeffects.

**Comment by lpw25** [comment]
Are you saying that flambda2 is trying to do unboxing based solely on the effects/coeffects of the operations, rather than based on knowing which operations are boxing operations?

**Comment by lthls** [comment]
> Are you saying that flambda2 is trying to do unboxing based solely on the effects/coeffects of the operations, rather than based on knowing which operations are boxing operations?

Not flambda2, just the classic mode. And yes, we don't do unboxing in classic mode, we just move operations around and hope that allocations and accesses cancel each other.

**Comment by lthls** [comment]
We are a bit more aggressive for boxing operations than for other kind of primitives, but that's all

**Comment by lpw25** [comment]
Right, so it should be safe to mark local allocations as `Delay`, and then you should get the same behaviour as cmmgen. General code motion operations are not safe for local allocations, and those should default to treating them as coeffectful (with regions as effectful). But the specific case of substitution is safe and so they can still be marked as `Delay`.
